### PR TITLE
feature(permissions): add embargo check to NonDegreePermissionStrategy

### PR DESCRIPTION
**Pull request by torbjokv**

### CODE REVIEW

File: publication-commons/src/main/java/no/unit/nva/publication/permission/strategy/PermissionStrategy.java [hunk 1]
```diff
@@ -45,6 +46,17 @@ protected boolean isDegree() {
                    .orElse(false);
     }
 
+    protected boolean isEmbargoDegree() {
+        return publication.getAssociatedArtifacts().stream()
```
Comment (joachimjorgensen) [inline publication-commons/src/main/java/no/unit/nva/publication/permission/strategy/PermissionStrategy.java:50]: Have to add a check for if the publication is a degree or not. A "normal" publication might also have embargo.

Comment (joachimjorgensen) [reply publication-commons/src/main/java/no/unit/nva/publication/permission/strategy/PermissionStrategy.java:50]:
```suggestion
        return isDegree() && publication.getAssociatedArtifacts().stream()
```

Comment (joachimjorgensen) [reply publication-commons/src/main/java/no/unit/nva/publication/permission/strategy/PermissionStrategy.java:50]: Think this should do the trick?

File: publication-commons/src/test/java/no/unit/nva/publication/permission/strategy/NonDegreePermissionStrategyTest.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+                                                cristinId, publication.getResourceOwner().getOwnerAffiliation());
+        var userInstance = RequestUtil.createUserInstanceFromRequest(requestInfo, identityServiceClient);
+
+        Assertions.assertTrue(PublicationPermissionStrategy
+                                  .create(publication, userInstance, uriRetriever)
+                                  .allowsAction(operation));
+    }
+
+    @Test
+    void shouldDenyCuratorWithMissingEmbargoAccess()
```
Comment (joachimjorgensen) [inline publication-commons/src/test/java/no/unit/nva/publication/permission/strategy/NonDegreePermissionStrategyTest.java:73]:
```suggestion
    void shouldDenyNonEmbargoThesisCuratorOnDegreeWithEmbargo()
```

File: publication-commons/src/test/java/no/unit/nva/publication/permission/strategy/NonDegreePermissionStrategyTest.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+                                                cristinId, publication.getResourceOwner().getOwnerAffiliation());
+        var userInstance = RequestUtil.createUserInstanceFromRequest(requestInfo, identityServiceClient);
+
+        Assertions.assertFalse(PublicationPermissionStrategy
+                                  .create(publication, userInstance, uriRetriever)
+                                  .allowsAction(PublicationOperation.UPDATE));
+    }
+
+    @Test
+    void shouldAllowCuratorWithEmbargoAccess()
```
Comment (joachimjorgensen) [inline publication-commons/src/test/java/no/unit/nva/publication/permission/strategy/NonDegreePermissionStrategyTest.java:96]:
```suggestion
    void shouldAllowEmbargoThesisCuratorOnDegreeWithEmbargo()
```

File: publication-commons/src/test/java/no/unit/nva/publication/permission/strategy/NonDegreePermissionStrategyTest.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+        Assertions.assertFalse(PublicationPermissionStrategy
+                                   .create(publication, userInstance, uriRetriever)
+                                   .allowsAction(operation));
+    }
+
+    @ParameterizedTest(name = "Should allow Curator {0} operation on degree resources belonging to the institution "
+                              + "with MANAGE_DEGREE access rights")
+    @EnumSource(value = PublicationOperation.class, mode = Mode.EXCLUDE, names = {"DELETE", "TERMINATE",
+        "TICKET_PUBLISH"})
+    void shouldAllowCuratorOnDegree(PublicationOperation operation)
```
Comment (joachimjorgensen) [inline publication-commons/src/test/java/no/unit/nva/publication/permission/strategy/NonDegreePermissionStrategyTest.java:52]:
```suggestion
    void shouldAllowThesisCuratorOnDegree(PublicationOperation operation)
```

File: publication-commons/src/test/java/no/unit/nva/publication/permission/strategy/NonDegreePermissionStrategyTest.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        throws JsonProcessingException, UnauthorizedException {
+
+        var institution = randomUri();
+        var resourceOwner = randomString();
+        var curatorUsername = randomString();
+        var cristinId = randomUri();
+
+        var publication = createDegreePhd(resourceOwner, institution, randomUri()).copy()
+                              .withStatus(PUBLISHED)
+                              .withAssociatedArtifacts(List.of(randomFileWithEmbargo()))
```
Comment (joachimjorgensen) [inline publication-commons/src/test/java/no/unit/nva/publication/permission/strategy/NonDegreePermissionStrategyTest.java:83]:
```suggestion
                              .withAssociatedArtifacts(List.of(randomFileWithEmbargo(), PublishedFileGenerator.random()))
```

File: publication-commons/src/test/java/no/unit/nva/publication/permission/strategy/NonDegreePermissionStrategyTest.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+                              .build();
+
+        var requestInfo = createUserRequestInfo(curatorUsername, institution, getCuratorAccessRightsWithDegree(),
+                                                cristinId, publication.getResourceOwner().getOwnerAffiliation());
+        var userInstance = RequestUtil.createUserInstanceFromRequest(requestInfo, identityServiceClient);
+
+        Assertions.assertFalse(PublicationPermissionStrategy
+                                  .create(publication, userInstance, uriRetriever)
+                                  .allowsAction(PublicationOperation.UPDATE));
+    }
```
Comment (joachimjorgensen) [inline publication-commons/src/test/java/no/unit/nva/publication/permission/strategy/NonDegreePermissionStrategyTest.java:139]: Duplicate of the test above as far as I can see. Only difference being the addition of a non embargo file in associated artifacts. Adding this file to the test on line 74 would give as much cover I would presume?

```suggestion
```

File: publication-commons/src/test/java/no/unit/nva/publication/permission/strategy/PublicationPermissionStrategyTest.java [hunk 7]
```diff
@@ -648,6 +659,15 @@ protected List<AccessRight> getCuratorAccessRightsWithDegree() {
         return accessRights;
     }
 
+    protected List<AccessRight> getCuratorAccessRightsWithEmbargoDegree() {
```
Comment (joachimjorgensen) [inline publication-commons/src/test/java/no/unit/nva/publication/permission/strategy/PublicationPermissionStrategyTest.java:662]: Kind of misleading name. Thought it meant returning all access rights with embargo degree, which would only be `MANAGE_DEGREE_EMBARGO`. 

Kind of risky to hard code these rights tho, as the single source of truth should be in identity service. But that might be a hassle.

```suggestion
    protected List<AccessRight> getAccessRightsForEmbargoThesisCurator() {
```

Comment (torbjokv) [reply publication-commons/src/test/java/no/unit/nva/publication/permission/strategy/PublicationPermissionStrategyTest.java]: yup, as long as identity-service owns the role definitions, doing something fancy is probably not worth it. We could maybe think about what should be included or not here

File: publication-commons/src/test/java/no/unit/nva/publication/permission/strategy/PublicationPermissionStrategyTest.java [hunk 8]
```diff
@@ -659,7 +659,7 @@ protected List<AccessRight> getCuratorAccessRightsWithDegree() {
         return accessRights;
     }
 
-    protected List<AccessRight> getCuratorAccessRightsWithEmbargoDegree() {
+    protected List<AccessRight> getAccessRightsForEmbargoThesisCurator() {
```
Comment (joachimjorgensen) [inline publication-commons/src/test/java/no/unit/nva/publication/permission/strategy/PublicationPermissionStrategyTest.java:662]: Eeeeegentlig burde de over også skifte navn. Men det får være opp til deg :)

**Review by joachimjorgensen — APPROVED** [review]

File: publication-commons/src/test/java/no/unit/nva/publication/permission/strategy/PublicationPermissionStrategyTest.java (hunk 8 above)
Comment (torbjokv) [reply publication-commons/src/test/java/no/unit/nva/publication/permission/strategy/PublicationPermissionStrategyTest.java:662]: your wish is my command! :-)

**Pull request merged**